### PR TITLE
Buffer implementation for makedrive

### PR DESCRIPTION
**Pull request by alicoding**

### CODE REVIEW

**Comment by alicoding** [comment]
@humphd this will obviously fail because I didn't have @gideonthomas adler's algo merged! I feel so derp right now LOL

File: client/src/client-server-sync.js [hunk 1]
```diff
@@ -1,13 +1,9 @@
 var syncID,
-    rsync = require('./rsync'),
+    rsync = require('../../lib/rsync'),
+    rsyncOptions = require( "../../lib/constants" ),
```
Comment (humphd) [inline client/src/client-server-sync.js]: You need `.rsyncDefaults`

Comment (humphd) [inline client/src/client-server-sync.js]: Don't mix quote types.  Stay with single-quotes to match the file you're in.

File: server/lib/sync.js [hunk 2]
```diff
@@ -74,36 +70,23 @@ function createError(code, message) {
...
         for (k = 0; k < diffs[i].contents[j].diffs.length; k++) {
-          if (Object.prototype.toString.call(diffs[i].contents[j].diffs[k].data) === "[object Uint8Array]") {
-            diffs[i].contents[j].diffs[k].data = {
-              __isUint8Array: true,
-              __array: u8toArray(diffs[i].contents[j].diffs[k].data)
-            };
+          buf = diffs[i].contents[j].diffs[k].data;
+          if (Buffer.isBuffer(buf)) {
+            diffs[i].contents[j].diffs[k].data = buf.toJSON();
           }
```
Comment (humphd) [inline server/lib/sync.js]: I hate these deep loops.  Why not change it to something like this? (NOTE: I haven't tested this)

``` javascript
function processDiff(diff) {
  if(diff.contents) {
    diff.contents.forEach(processDiff);
  } else {
    if(Buffer.isBuffer(diff.data)) {
      diff.data = diff.data.toJSON();
    }
  }
}

diffs.forEach(processDiff);
```

File: client/src/comms.js [hunk 3]
```diff
@@ -27,26 +20,20 @@ var syncState;
...
       for ( k = 0; k < diffs[i].diffs.length; k++ ) {
         if ( diffs[i].diffs[k].data ) {
-          diffs[i].diffs[k].data = diffs[i].diffs[k].data;
-          if ( diffs[i].diffs[k].data.__isUint8Array ) {
-            diffs[i].diffs[k].data = new Uint8Array( diffs[i].diffs[k].data.__array );
-          }
+          diffs[i].diffs[k].data = new Buffer(diffs[i].diffs[k].data);
         }
       }
     }
```
Comment (humphd) [inline client/src/comms.js]: Same thing here, let's simplify:

I hate these deep loops.  Why not change it to something like this? (NOTE: I haven't tested this)

``` javascript
function processDiff(diff) {
  if(diff.contents) {
    diff.contents.forEach(processDiff);
  } else {
    if(Buffer.isBuffer(diff.data)) {
      diff.data = new Buffer(diff.data);
    }
  }
}

diffs.forEach(processDiff);
```

File: tests/lib/util.js [hunk 4]
```diff
@@ -56,8 +58,7 @@ if(!uploadFound) {
         name: username
       });
 
-      // XXXhumph: this hack is just here until we switch Filer to Buffers
-      fileData = new Uint8Array(fileData.toJSON());
+      fileData = new Buffer(fileData.toJSON());
```
Comment (humphd) [inline tests/lib/util.js]: It looks to me like you're turning a buffer into json so you can turn it back into a buffer.  Why not just remove this line completely?

**Comment by humphd** [comment]
You can further generalize the processDiff stuff so it takes a function that tells it how to process the data:

``` javascript
// new file at: lib/diff.js
function processDiff(diff, processDataFn) {
  if(diff.contents) {
    diff.contents.forEach(processDiff);
  } else {
    if(Buffer.isBuffer(diff.data)) {
      diff.data = processDataFn(diff.data)
    }
  }
}

module.exports.serialize = function(diffs) {
  diffs.forEach(function(diff){
    processDiff(diff, function bufferToJSON(data) { return data.toJSON(); });
  });
};

module.exports.deserialize = function(diffs) {
  diffs.forEach(function(diff){
    processDiff(diff, function jsonToBuffer(data) { new Buffer(data); });
  });
};
```

File: tests/lib/util.js (hunk 4 above)
Comment (alicoding) [inline tests/lib/util.js]: I did removed and forgot to push this up haha. Good catch!

File: client/src/comms.js [hunk 5]
```diff
@@ -1,18 +1,12 @@
...
     uri,
-    checksum = {};
-
-var rsyncOptions = {
-  time: true,
-  recursive: true,
-  size: 5
-};
+    rsyncOptions = require('../../lib/constants').rsyncDefaults,
+    processDiff = require('../../lib/diff').deserialize;
```
Comment (humphd) [inline client/src/comms.js]: The naming here is odd.  If you only want the deserialize bit, I'd name it `deserializeDiff` vs. `processDiff` for clarity.

File: lib/diff.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+module.exports.serialize = function(diffs) {
+  if(!diffs.length) {
+    return diffs;
+  }
+  diffs.forEach(function(diff){
+    processDiff(diff, function bufferToJSON(data) { return data.toJSON(); });
+  });
+};
```
Comment (humphd) [inline lib/diff.js:18]: You should `return diffs` here, since your caller expects it.

File: lib/diff.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+module.exports.deserialize = function(diffs) {
+  if(!diffs.length) {
+    return diffs;
+  }
+  diffs.forEach(function(diff){
+    processDiff(diff, function jsonToBuffer(data) { new Buffer(data); });
+  });
+};
```
Comment (humphd) [inline lib/diff.js:28]: Same here.

File: client/src/comms.js [hunk 8]
```diff
@@ -136,7 +101,7 @@ module.exports = {
                   var diffs = data.content.diffs;
                   path = data.content.path;
 
-                  convert.call( diffs );
+                  diffs = processDiff( diffs );
```
Comment (humphd) [inline client/src/comms.js]: Here you're expecting the diffs object to come back.  Your module only returns it in one case.  It needs to happen in every case.

File: server/lib/sync.js [hunk 9]
```diff
@@ -7,9 +7,12 @@ var env = require( "./environment" ),
     filesystem = require( "./filesystem" ),
     uuid = require( "node-uuid" ),
     emitter = new ( require( "events" ).EventEmitter )(),
+    Buffer = require('filer').Buffer,
     SyncMessage = require( "./syncmessage" );
 
-var rsync = require( "./rsync" );
+var rsync = require( "../../lib/rsync" ),
+    processDiff = require('../../lib/diff').serialize,
```
Comment (humphd) [inline server/lib/sync.js]: s/processDiff/serializeDiff/

File: server/routes/index.js [hunk 10]
```diff
@@ -181,20 +182,20 @@ module.exports = function createRoutes( app, webmakerAuth  ) {
         form.handlePart(part);
       } else {
         part.addListener('data', function(data) {
-          // Parse JSON diffs to Uint8Array
+          // Parse JSON diffs to Buffer
```
Comment (humphd) [inline server/routes/index.js:185]: We should reuse the lib/diff code here too, and allow a processDiffFn to get passed in, where you just return data.

Comment (alicoding) [inline server/routes/index.js:185]: We're going to rip this off in next patch, so should I do it right now?

File: client/src/comms.js [hunk 11]
```diff
@@ -1,18 +1,12 @@
...
     uri,
-    checksum = {};
-
-var rsyncOptions = {
-  time: true,
-  recursive: true,
-  size: 5
-};
+    rsyncOptions = require('../../lib/constants').rsyncDefaults,
+    deSerializeDiff = require('../../lib/diff').deserialize;
```
Comment (humphd) [inline client/src/comms.js]: lower case s please--`deserializeDiff`

File: lib/diff.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
+// new file at: lib/diff.js
```
Comment (humphd) [inline lib/diff.js]: Remove this comment :)

Comment (alicoding) [inline lib/diff.js]: Oops!

File: server/routes/index.js (hunk 10 above)
Comment (humphd) [inline server/routes/index.js:185]: Leave it then.

**Comment by humphd** [comment]
r=me with those nits fixed.

**Pull request merged**